File: packages/react-native/ios/UpdateStore.swift
import Foundation

/**
 <Application Support>/open-ota/
   slots/A, slots/B   extracted bundles (the `expo export` layout, preserved)
   state.json         which slot is what
   events.jsonl       events produced natively (rollbacks), drained by JS
   tmp/               in-flight downloads

 Every write is atomic: a half-written state.json read at boot would be
 indistinguishable from "no update installed" and would strand the device on the
 embedded bundle.
 */
struct OtaState: Codable {
  var currentReleaseId: String?
  var currentSlot: String?
  var currentLabel: Int?
  var previousReleaseId: String?
  var previousSlot: String?
  var previousLabel: Int?
  var pendingReleaseId: String?
  var pendingSlot: String?
  var pendingLabel: Int?
  var pendingVerification = false
  var failedReleaseIds: [String] = []
  var previewReleaseId: String?
  var deviceId: String?
  var channelOverride: String?

  init() {}

  init(from decoder: Decoder) throws {
    let values = try decoder.container(keyedBy: CodingKeys.self)
    currentReleaseId = try values.decodeIfPresent(String.self, forKey: .currentReleaseId)
    currentSlot = try values.decodeIfPresent(String.self, forKey: .currentSlot)
    currentLabel = try values.decodeIfPresent(Int.self, forKey: .currentLabel)
    previousReleaseId = try values.decodeIfPresent(String.self, forKey: .previousReleaseId)
    previousSlot = try values.decodeIfPresent(String.self, forKey: .previousSlot)
    previousLabel = try values.decodeIfPresent(Int.self, forKey: .previousLabel)
    pendingReleaseId = try values.decodeIfPresent(String.self, forKey: .pendingReleaseId)
    pendingSlot = try values.decodeIfPresent(String.self, forKey: .pendingSlot)
    pendingLabel = try values.decodeIfPresent(Int.self, forKey: .pendingLabel)
    pendingVerification = try values.decodeIfPresent(Bool.self, forKey: .pendingVerification) ?? false
    failedReleaseIds = try values.decodeIfPresent([String].self, forKey: .failedReleaseIds) ?? []
    previewReleaseId = try values.decodeIfPresent(String.self, forKey: .previewReleaseId)
    deviceId = try values.decodeIfPresent(String.self, forKey: .deviceId)
    channelOverride = try values.decodeIfPresent(String.self, forKey: .channelOverride)
  }
}

final class UpdateStore {

  static let slotA = "A"
  static let slotB = "B"

  let directory: URL
  let tmpDirectory: URL
  private let stateURL: URL
  private let eventsURL: URL

  var state: OtaState

  init() {
    let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
      ?? URL(fileURLWithPath: NSTemporaryDirectory())
    directory = base.appendingPathComponent("open-ota", isDirectory: true)
    tmpDirectory = directory.appendingPathComponent("tmp", isDirectory: true)
    stateURL = directory.appendingPathComponent("state.json")
    eventsURL = directory.appendingPathComponent("events.jsonl")

    var isDirectory: ObjCBool = false
    if !FileManager.default.fileExists(atPath: directory.path, isDirectory: &isDirectory) {
      try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
      var resourceValues = URLResourceValues()
      resourceValues.isExcludedFromBackup = true
      var mutable = directory
      try? mutable.setResourceValues(resourceValues)
    }

    if let data = try? Data(contentsOf: stateURL),
       let decoded = try? JSONDecoder().decode(OtaState.self, from: data) {
      state = decoded
    } else {
      state = OtaState()
    }
  }

  func slotDirectory(_ slot: String) -> URL {
    directory.appendingPathComponent("slots", isDirectory: true)
      .appendingPathComponent(slot, isDirectory: true)
  }

  /// The slot not backing the running bundle. Whatever lived there is forfeit.
  func freeSlot() -> String {
    state.currentSlot == UpdateStore.slotA ? UpdateStore.slotB : UpdateStore.slotA
  }

  func write() throws {
    try FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    let data = try JSONEncoder().encode(state)
    try data.write(to: stateURL, options: .atomic)
  }

  /* ---------------------------------------------------------------- telemetry */

  /// Natively generated events (crash rollbacks) outlive the process that saw them.
  func appendEvent(_ event: [String: Any]) {
    guard let data = try? JSONSerialization.data(withJSONObject: event),
          var line = String(data: data, encoding: .utf8)
    else { return }
    line += "\n"

    try? FileManager.default.createDirectory(at: directory, withIntermediateDirectories: true)
    if let handle = try? FileHandle(forWritingTo: eventsURL) {
      defer { try? handle.close() }
      _ = try? handle.seekToEnd()
      try? handle.write(contentsOf: Data(line.utf8))
    } else {
      try? Data(line.utf8).write(to: eventsURL, options: .atomic)
    }
  }

  // ponytail: read-then-delete, so a crash between the two loses a counter.
  // These are operational counters, not billing.
  func takeEvents() -> [[String: Any]] {
    guard let text = try? String(contentsOf: eventsURL, encoding: .utf8) else { return [] }
    try? FileManager.default.removeItem(at: eventsURL)
    return text.split(separator: "\n").compactMap { line in
      guard let data = line.data(using: .utf8) else { return nil }
      return (try? JSONSerialization.jsonObject(with: data)) as? [String: Any]
    }
  }
}

/// Typed failures; the code travels to JS as the promise rejection code.
struct OtaError: Error {
  let code: String
  let message: String

  init(_ code: String, _ message: String) {
    self.code = code
    self.message = message
  }

  static let notConfigured = "ERR_OTA_NOT_CONFIGURED"
  static let signatureInvalid = "ERR_OTA_SIGNATURE_INVALID"
  static let manifestMismatch = "ERR_OTA_MANIFEST_MISMATCH"
  static let downloadFailed = "ERR_OTA_DOWNLOAD_FAILED"
  static let hashMismatch = "ERR_OTA_HASH_MISMATCH"
  static let extractFailed = "ERR_OTA_EXTRACT_FAILED"
  static let nothingPending = "ERR_OTA_NOTHING_PENDING"
  static let stateWriteFailed = "ERR_OTA_STATE_WRITE_FAILED"
  static let previewInvalid = "ERR_OTA_PREVIEW_INVALID"
}

File: packages/react-native/android/src/main/java/dev/openota/UpdateStore.kt
package dev.openota

import org.json.JSONArray
import org.json.JSONObject
import java.io.File
import java.io.FileOutputStream

/**
 * <app-data>/open-ota/
 *   slots/A, slots/B   extracted bundles (the `expo export` layout, preserved)
 *   state.json         which slot is what
 *   events.jsonl       events produced natively (rollbacks), drained by JS
 *   tmp/               in-flight downloads
 *
 * Every write goes through a temp file + rename: a half-written state.json read
 * at boot would be indistinguishable from "no update installed" and would strand
 * the device on the embedded bundle.
 */
internal class UpdateStore(root: File) {

  data class State(
    var currentReleaseId: String? = null,
    var currentSlot: String? = null,
    var currentLabel: Int? = null,
    var previousReleaseId: String? = null,
    var previousSlot: String? = null,
    var previousLabel: Int? = null,
    var pendingReleaseId: String? = null,
    var pendingSlot: String? = null,
    var pendingLabel: Int? = null,
    var pendingVerification: Boolean = false,
    var failedReleaseIds: MutableList<String> = mutableListOf(),
    var previewReleaseId: String? = null,
    var deviceId: String? = null,
    var channelOverride: String? = null,
  ) {
    fun snapshot(): State = copy(failedReleaseIds = failedReleaseIds.toMutableList())
  }

  val dir: File = File(root, "open-ota")
  private val stateFile = File(dir, "state.json")
  private val eventsFile = File(dir, "events.jsonl")
  val tmpDir: File = File(dir, "tmp")

  val state: State by lazy { read() }

  fun slotDir(slot: String): File = File(File(dir, "slots"), slot)

  /** The slot not backing the running bundle. Whatever lived there is forfeit. */
  fun freeSlot(): String = if (state.currentSlot == SLOT_A) SLOT_B else SLOT_A

  fun write(value: State = state) {
    dir.mkdirs()
    val json = JSONObject()
      .putOpt("currentReleaseId", value.currentReleaseId)
      .putOpt("currentSlot", value.currentSlot)
      .putOpt("currentLabel", value.currentLabel)
      .putOpt("previousReleaseId", value.previousReleaseId)
      .putOpt("previousSlot", value.previousSlot)
      .putOpt("previousLabel", value.previousLabel)
      .putOpt("pendingReleaseId", value.pendingReleaseId)
      .putOpt("pendingSlot", value.pendingSlot)
      .putOpt("pendingLabel", value.pendingLabel)
      .put("pendingVerification", value.pendingVerification)
      .put("failedReleaseIds", JSONArray(value.failedReleaseIds))
      .putOpt("previewReleaseId", value.previewReleaseId)
      .putOpt("deviceId", value.deviceId)
      .putOpt("channelOverride", value.channelOverride)

    val tmp = File(dir, "state.json.tmp")
    FileOutputStream(tmp).use {
      it.write(json.toString().toByteArray(Charsets.UTF_8))
      it.fd.sync()
    }
    if (!tmp.renameTo(stateFile)) {
      tmp.delete()
      throw OtaError(OtaError.STATE_WRITE_FAILED, "could not persist ${stateFile.path}")
    }
  }

  /** Replaces the in-memory state in place — callers hold a reference to it. */
  fun restore(snapshot: State) {
    state.apply {
      currentReleaseId = snapshot.currentReleaseId
      currentSlot = snapshot.currentSlot
      currentLabel = snapshot.currentLabel
      previousReleaseId = snapshot.previousReleaseId
      previousSlot = snapshot.previousSlot
      previousLabel = snapshot.previousLabel
      pendingReleaseId = snapshot.pendingReleaseId
      pendingSlot = snapshot.pendingSlot
      pendingLabel = snapshot.pendingLabel
      pendingVerification = snapshot.pendingVerification
      failedReleaseIds = snapshot.failedReleaseIds.toMutableList()
      previewReleaseId = snapshot.previewReleaseId
      deviceId = snapshot.deviceId
      channelOverride = snapshot.channelOverride
    }
    write()
  }

  private fun read(): State {
    val raw = runCatching { stateFile.readText() }.getOrNull() ?: return State()
    val json = runCatching { JSONObject(raw) }.getOrNull() ?: return State()
    val failed = json.optJSONArray("failedReleaseIds")
    return State(
      currentReleaseId = json.str("currentReleaseId"),
      currentSlot = json.str("currentSlot"),
      currentLabel = json.int("currentLabel"),
      previousReleaseId = json.str("previousReleaseId"),
      previousSlot = json.str("previousSlot"),
      previousLabel = json.int("previousLabel"),
      pendingReleaseId = json.str("pendingReleaseId"),
      pendingSlot = json.str("pendingSlot"),
      pendingLabel = json.int("pendingLabel"),
      pendingVerification = json.optBoolean("pendingVerification", false),
      failedReleaseIds = MutableList(failed?.length() ?: 0) { failed!!.optString(it) },
      previewReleaseId = json.str("previewReleaseId"),
      deviceId = json.str("deviceId"),
      channelOverride = json.str("channelOverride"),
    )
  }

  /* -------------------------------------------------------------- telemetry */

  /** Natively generated events (crash rollbacks) outlive the process that saw them. */
  fun appendEvent(event: JSONObject) {
    runCatching {
      dir.mkdirs()
      FileOutputStream(eventsFile, true).use {
        it.write((event.toString() + "\n").toByteArray(Charsets.UTF_8))
        it.fd.sync()
      }
    }
  }

  // ponytail: read-then-delete, so a crash between the two loses a counter.
  // These are operational counters, not billing.
  fun takeEvents(): List<Map<String, Any?>> {
    val text = runCatching { eventsFile.readText() }.getOrNull() ?: return emptyList()
    eventsFile.delete()
    return text.lineSequence()
      .filter { it.isNotBlank() }
      .mapNotNull { line -> runCatching { JSONObject(line).toMap() }.getOrNull() }
      .toList()
  }

  companion object {
    const val SLOT_A = "A"
    const val SLOT_B = "B"

    private fun JSONObject.str(key: String): String? =
      if (isNull(key)) null else optString(key).ifEmpty { null }

    private fun JSONObject.int(key: String): Int? = if (isNull(key)) null else optInt(key)

    private fun JSONObject.toMap(): Map<String, Any?> = keys().asSequence().associateWith { key ->
      when (val v = get(key)) {
        is JSONObject -> v.toMap()
        JSONObject.NULL -> null
        else -> v
      }
    }
  }
}

/** Typed failures; the code travels to JS as the promise rejection code. */
internal class OtaError(val code: String, message: String, cause: Throwable? = null) :
  Exception(message, cause) {
  companion object {
    const val NOT_CONFIGURED = "ERR_OTA_NOT_CONFIGURED"
    const val SIGNATURE_INVALID = "ERR_OTA_SIGNATURE_INVALID"
    const val MANIFEST_MISMATCH = "ERR_OTA_MANIFEST_MISMATCH"
    const val DOWNLOAD_FAILED = "ERR_OTA_DOWNLOAD_FAILED"
    const val HASH_MISMATCH = "ERR_OTA_HASH_MISMATCH"
    const val EXTRACT_FAILED = "ERR_OTA_EXTRACT_FAILED"
    const val NOTHING_PENDING = "ERR_OTA_NOTHING_PENDING"
    const val RELOAD_UNSUPPORTED = "ERR_OTA_RELOAD_UNSUPPORTED"
    const val STATE_WRITE_FAILED = "ERR_OTA_STATE_WRITE_FAILED"
    const val PREVIEW_INVALID = "ERR_OTA_PREVIEW_INVALID"
  }
}
